<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Подготовка-данных.-Первый-взгляд" data-toc-modified-id="Подготовка-данных.-Первый-взгляд-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подготовка данных. Первый взгляд</a></span></li><li><span><a href="#Подготовка-данных.-Обработка-пропусков" data-toc-modified-id="Подготовка-данных.-Обработка-пропусков-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка данных. Обработка пропусков</a></span></li><li><span><a href="#Подготовка-данных.-Шлифуем-данные" data-toc-modified-id="Подготовка-данных.-Шлифуем-данные-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Подготовка данных. Шлифуем данные</a></span></li><li><span><a href="#Подготовка-данных.-Признаки-и-раделение-выборок" data-toc-modified-id="Подготовка-данных.-Признаки-и-раделение-выборок-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Подготовка данных. Признаки и раделение выборок</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Исследование-задачи.-Проверка-сбалансированности" data-toc-modified-id="Исследование-задачи.-Проверка-сбалансированности-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Исследование задачи. Проверка сбалансированности</a></span></li><li><span><a href="#Исследование-задачи.-Обучение-моделей-без-балансировки" data-toc-modified-id="Исследование-задачи.-Обучение-моделей-без-балансировки-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Исследование задачи. Обучение моделей без балансировки</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Борьба-с-дисбаллансом." data-toc-modified-id="Борьба-с-дисбаллансом.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Борьба с дисбаллансом.</a></span></li><li><span><a href="#Борьба-с-дисбаллансом.-Обучение-модели-на-сбалансированной-выборке" data-toc-modified-id="Борьба-с-дисбаллансом.-Обучение-модели-на-сбалансированной-выборке-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Борьба с дисбаллансом. Обучение модели на сбалансированной выборке</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

In [1]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.utils import shuffle
from sklearn.compose import ColumnTransformer

from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import confusion_matrix

import sklearn
import numpy as np

## Подготовка данных

### Подготовка данных. Первый взгляд

In [2]:
data = pd.read_csv('/datasets/Churn.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


**Вывод 1.1**
* "Данные готовы для первичного анализа, но мы видим пропуски в столбце Tenure: скорее всего это связано с тем, что человек стал клиентом банка только в текущем году, поэтому заменим пропуски на нули в следующей ячейке"
* Второй вариант - посчитаю % пропусков от общего кол-ва, и исходя из рез-ов, придумаю решение этой проблемы.

### Подготовка данных. Обработка пропусков

In [3]:
#data.loc[data['Tenure'].isna(), 'Tenure'] = 0.0
print(data.Tenure.isna().sum() / data.Tenure.notna().sum() * 100, '%')

9.99890001099989 %


In [4]:
tenure_nan = data.loc[data.Tenure.isna()]
#tenure_nan.head()
data = data.dropna(subset=['Tenure']).reset_index(drop=True)

In [5]:
data.head(10)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   
5     8.0  113755.78              2          1               0   
6     7.0       0.00              2          1               1   
7     4.0  115046.74              4          1               0   
8     4.0  142051.07              2          0               1   
9     2.0  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

**Вывод 1.2**
* Итак, данные в порядке, теперь нам осталось подготовить для машинного обучения:
1. Нужно переместить стобецы с фамилиями и id, RowNumber клиентов в отдельный датафрейм, так как они нам не понадобяться.
2. Применить метод OHE для того, чтобы не было проблем с категориальными значениями
3. Выделим количественные признаки, для упрощения дальнейшей работы и сохраним их в списке `numeric`

### Подготовка данных. Шлифуем данные

In [6]:
clients_id = data[['CustomerId', 'Surname', 'RowNumber']]
#clients_id['Surname'] = data['Surname']
data = data.drop(['CustomerId', 'Surname', 'RowNumber'], axis=1)
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [7]:
numeric = ['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'EstimatedSalary']

In [8]:
#data['Geography'].unique()
#clients_id.head()

In [9]:
#data_df = pd.get_dummies(data_df, drop_first=True)
#Новый вариант с OHE в следующей ячейке

In [10]:
categories = ['Geography', 'Gender']
ohe = OneHotEncoder(sparse=False)
#ohe.fit(data[categories])

#sklearn.__version__

In [11]:
def get_enc_fit(df):
    temp_df = pd.DataFrame(data=enc.fit_transform(df[categories]), columns=categories)
    df.drop(columns=categories, axis=1, inplace=True)
    df = pd.concat([df.reset_index(drop=True), temp_df], axis=1)
    return df
#data = get_ohe(data)

In [12]:
def get_enc(df):
    temp_df = pd.DataFrame(data=enc.transform(df[categories]), columns=categories)
    df.drop(columns=categories, axis=1, inplace=True)
    df = pd.concat([df.reset_index(drop=True), temp_df], axis=1)
    return df
#data = get_ohe(data)

In [13]:
def get_ohe_fit(df):
    temp_df = pd.DataFrame(data=ohe.fit_transform(df[categories]).toarray(), columns=ohe.get_feature_names())
    df.drop(columns=categories, axis=1, inplace=True)
    df = pd.concat([df.reset_index(drop=True), temp_df], axis=1)
    return df
#data = get_ohe(data)

In [14]:
def get_ohe(df):
    temp_df = pd.DataFrame(data=ohe.transform(df[categories]).toarray(), columns=ohe.get_feature_names())
    df.drop(columns=categories, axis=1, inplace=True)
    df = pd.concat([df.reset_index(drop=True), temp_df], axis=1)
    return df
#data = get_ohe(data)

In [15]:
data.head(10)

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42     2.0       0.00              1   
1          608     Spain  Female   41     1.0   83807.86              1   
2          502    France  Female   42     8.0  159660.80              3   
3          699    France  Female   39     1.0       0.00              2   
4          850     Spain  Female   43     2.0  125510.82              1   
5          645     Spain    Male   44     8.0  113755.78              2   
6          822    France    Male   50     7.0       0.00              2   
7          376   Germany  Female   29     4.0  115046.74              4   
8          501    France    Male   44     4.0  142051.07              2   
9          684    France    Male   27     2.0  134603.88              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0  
5          1               0        149756.71       1  
6          1               1         10062.80       0  
7          1               0        119346.88       1  
8          0               1         74940.50       0  
9          1               1         71725.73       0

**Вывод 1.3**
- Теперь у нас данные готовы для машинного обучения, они хранятся в `data`
- B `clients_id` хранятся данные о том, какой id кому принадлежит.
- Далее подготовим признаки.

### Подготовка данных. Признаки и раделение выборок 

In [16]:
target = data['Exited']
features = data.drop(['Exited'], axis=1)
#print(features.shape, target.shape)
#Далее разделим их на выборки
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.6, random_state=12345, stratify=target)
features_valid, features_test, target_valid, target_test = train_test_split(
    features_valid, target_valid, test_size=0.5, random_state=12345, stratify=target_valid)

#print(features_train.shape, target_train.shape)

In [17]:
#enc = OrdinalEncoder()
ohe = OneHotEncoder(drop='first')
#ohe.fit(features_train)
features_train_enc = get_ohe_fit(features_train)
features_valid_enc = get_ohe(features_valid)
features_test_enc = get_ohe(features_test)
target_train.reset_index(drop=True, inplace=True)
target_valid.reset_index(drop=True, inplace=True)
target_test.reset_index(drop=True, inplace=True)

/opt/conda/lib/python3.9/site-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [18]:
features_test_enc.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          850   61     3.0  141784.02              1          1   
1          616   43     3.0  120867.18              1          1   
2          497   44     2.0  121250.04              1          0   
3          655   72     5.0  138089.97              2          1   
4          429   60     7.0       0.00              2          1   

   IsActiveMember  EstimatedSalary  x0_Germany  x0_Spain  x1_Male  
0               1         92053.75         1.0       0.0      1.0  
1               0         18761.92         0.0       1.0      0.0  
2               1         79691.40         0.0       1.0      1.0  
3               1         99920.41         1.0       0.0      0.0  
4               1        163691.48         0.0       0.0      1.0

**Вывод 1.4**
- Признаки готовы, можно приступать к выполнению главной задачи

## Исследование задачи

### Исследование задачи. Проверка сбалансированности

In [19]:
features_zeros = features_train_enc[target_train == 0]
features_ones = features_train_enc[target_train == 1]

target_zeros = target_train[target_train == 0]
target_ones = target_train[target_train == 1]

print(features_zeros.shape[0])
print(features_ones.shape[0])
print(target_zeros.shape[0])
print(target_ones.shape[0])

2894
742
2894
742


**Вывод 2.1**: Отрицательный класс больше чем в 3 раза превышает положительный: испроавим в следующем шаге.

### Исследование задачи. Обучение моделей без балансировки

In [20]:
best_tree_depth = 0
best_tree_f1 = 0

In [21]:
for depth in range(1, 16):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_train_enc, target_train)

    predicted_valid = model.predict(features_valid_enc)
    probabilities_valid = model.predict_proba(features_valid_enc)
    probabilities_one_valid = probabilities_valid[:, 1]
    if f1_score(target_valid, predicted_valid) > best_tree_f1:
        best_tree_depth = depth
        best_tree_f1 = f1_score(target_valid, predicted_valid)
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

print(best_tree_f1)
print(best_tree_depth)
print(auc_roc)
print(confusion_matrix(target_valid, predicted_valid))

0.5567226890756302
7
0.8163350940620143
[[1900  271]
 [ 284  272]]


- Результаты показали, что модель хорошо справляется с распознаванием отрицательного класса, но дело плохо с положительным

In [22]:
model = LogisticRegression(solver='lbfgs', random_state=12345, class_weight='balanced')
model.fit(features_train_enc, target_train)

predicted_valid = model.predict(features_valid_enc)
probabilities_valid = model.predict_proba(features_valid_enc)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

print(f1_score(target_valid, predicted_valid))
print(auc_roc)
print(confusion_matrix(target_valid, predicted_valid))

0.43278301886792453
0.6958609068525925
[[1398  773]
 [ 189  367]]


- LR часто определяет отрицаительный класс, как отрицательный

In [23]:
best_forest_depth = 0
best_forest_f1 = 0

In [24]:
for depth in range(1, 16):
    model = RandomForestClassifier(max_depth=depth, n_estimators=40, random_state=12345)
    model.fit(features_train_enc, target_train)

    predicted_valid = model.predict(features_valid_enc)
    probabilities_valid = model.predict_proba(features_valid_enc)
    probabilities_one_valid = probabilities_valid[:, 1]
    if f1_score(target_valid, predicted_valid) > best_forest_f1:
        best_forest_f1 = f1_score(target_valid, predicted_valid)
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        best_forest_depth = depth

print(best_forest_f1)
print(best_forest_depth)
print(auc_roc)
print(confusion_matrix(target_valid, predicted_valid))

0.5479768786127168
11
0.8460958547763356
[[2092   79]
 [ 318  238]]


- У леса дела чуть лучше чем у дерева

**Вывод 2.2**: 
- Модели обученые на несбалансированной выборке показывают довольно плохие результаты. Но лучшим оказывается модель случайного леса, если судить по метрикам. (Также использовал аргумент class_weight='balanced', но результаты изменились незначительно)
- Без балансировки не получается достичь необходимого результата (f1 > 0.59) -> нужно балансировать

## Борьба с дисбалансом

### Борьба с дисбаллансом. 

- Для этого будем использовать увеличение выборки с положительным классом.

In [25]:
#Используем функцию
def upsample(features, target, repeat):
    repeat = repeat
    features_upsample = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsample = pd.concat([target_zeros] + [target_ones] * repeat)
    features_upsampled, target_upsampled = shuffle(features_upsample, target_upsample, random_state=12345)
    return features_upsampled, target_upsampled

- Теперь подберем оптимальный коэф `repeat` для `upsample()`

In [26]:
best_repeat = 1

In [27]:
features_train_enc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3636 entries, 0 to 3635
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      3636 non-null   int64  
 1   Age              3636 non-null   int64  
 2   Tenure           3636 non-null   float64
 3   Balance          3636 non-null   float64
 4   NumOfProducts    3636 non-null   int64  
 5   HasCrCard        3636 non-null   int64  
 6   IsActiveMember   3636 non-null   int64  
 7   EstimatedSalary  3636 non-null   float64
 8   x0_Germany       3636 non-null   float64
 9   x0_Spain         3636 non-null   float64
 10  x1_Male          3636 non-null   float64
dtypes: float64(6), int64(5)
memory usage: 312.6 KB


In [28]:
for repeat in range(1, 10):
    model = RandomForestClassifier(max_depth=best_forest_depth, n_estimators=60, random_state=12345)
    
    features_upsampled, target_upsampled = upsample(features_train_enc, target_train, repeat)
    model.fit(features_upsampled, target_upsampled)
    
    predicted_valid = model.predict(features_valid_enc)
    probabilities_valid = model.predict_proba(features_valid_enc)
    probabilities_one_valid = probabilities_valid[:, 1]
    if f1_score(target_valid, predicted_valid) > best_forest_f1:
        best_forest_f1 = f1_score(target_valid, predicted_valid)
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        best_repeat = repeat

print(best_forest_f1)
print(best_forest_depth)
print(auc_roc)
print(confusion_matrix(target_valid, predicted_valid))
print(best_repeat)

0.6110590440487348
11
0.8505578770516522
[[1774  397]
 [ 169  387]]
3


- Ориентируясь на значение f1 меры, выявил, что положительный класс нужно увеличить в 3 раза для оптимального обучения модели

In [29]:
features_upsampled, target_upsampled = upsample(features_train_enc, target_train, best_repeat)

In [30]:
print(features_upsampled.shape, target_upsampled.shape)

(5120, 11) (5120,)


**Вывод 3.1**: 
- Теперь отрицательный и положительный в тренировочных выборках сбалансированы для оптимальной работы модели
- Далее обучим случайный лес уже на сбалансированных выборках и посмотрим на результат

### Борьба с дисбаллансом. Обучение модели на сбалансированной выборке

In [31]:
best_depth = 0
best_f1 = 0
best_auc = 0.5

- Также Подберем оптимальные параметры глубины и кол-ва деревьев для максимизации f1

In [32]:
for depth in range(1, 16):
    model = RandomForestClassifier(max_depth=depth, n_estimators=40, random_state=12345)
    model.fit(features_upsampled, target_upsampled)

    predicted_valid = model.predict(features_valid_enc)
    probabilities_valid = model.predict_proba(features_valid_enc)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    if f1_score(target_valid, predicted_valid) > best_f1:
        best_f1 = f1_score(target_valid, predicted_valid)
        best_depth = depth
        best_auc = auc_roc

print(best_depth)
print(best_f1)
print(best_auc)

9
0.6170212765957447
0.8493400581239292


- Max_depth = 9
- Теперь вручную подберем кол-во деревьев

In [39]:
model = RandomForestClassifier(max_depth=best_depth, n_estimators=70, random_state=12345)
model.fit(features_upsampled, target_upsampled)

predicted_valid = model.predict(features_valid_enc)
probabilities_valid = model.predict_proba(features_valid_enc)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
f1 = f1_score(target_valid, predicted_valid)
print(f1, auc_roc)

0.6217430368373764 0.8523746640642345


**Вывод 3.2**: 
- Экспериментальным путем подобрал оптимальные параметры и гиперпараметры для увеличения выборки и обучения модели
- С помощью этого достиг необходимого значения F1 меры и максимизировал его.
- Лучшее значение f1_score: 0.62
- Лучшее значение auc_roc: 0.85

## Тестирование модели

In [34]:
predicted_test = model.predict(features_test_enc)
probabilities_valid = model.predict_proba(features_test_enc)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_valid)

print(f1_score(target_test, predicted_test))
print(auc_roc)

0.6185022026431718
0.862423735045113


## Общий вывод

- (п.1)Предобработка
    1. Сначала обработали данные: пропуски в столбце Tenure составили < 10%: удалил их
    2. Методом OHE преобразовали категориальные стобцы в численные
    3. Подготовил признаки и разделил их на тренировочную, тестовую, валидационную выборки в соотношении 3:1:1 
- (п.2)Исследование задачи и решение проблемы разбалансировки
    1. Отриуательный класс оказался почти в 4 раза больше положительного
    2. Обучил несколько моделей на несбалансированных выборках и понял, что алгоритмы показвают не достаточные значения f1 меры и auc_roc.
- (п.3)Борьба с дисбалансом
    1. Сбалансировал данные, увеличив двукратно положительный класс(перебирая разные коэф-ты и доводя значение f1 меры до значения > 0.59
    2. Подобрал оптимальные гиперпараметры для наилучшей модели (случайный лес) best_depth = 8, n_est = 50
- (п.4)Тестирование модели
    1. Наилучшая модель показала хорошие значения метрик `f1 = 0.618` и `auc_roc = 0.862`
Задачи проекта выполнены.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Выполнен шаг 1: данные подготовлены
- [ ]  Выполнен шаг 2: задача исследована
    - [ ]  Исследован баланс классов
    - [ ]  Изучены модели без учёта дисбаланса
    - [ ]  Написаны выводы по результатам исследования
- [ ]  Выполнен шаг 3: учтён дисбаланс
    - [ ]  Применено несколько способов борьбы с дисбалансом
    - [ ]  Написаны выводы по результатам исследования
- [ ]  Выполнен шаг 4: проведено тестирование
- [ ]  Удалось достичь *F1*-меры не менее 0.59
- [ ]  Исследована метрика *AUC-ROC*